lib/kino/smart_cell/chart_builder.ex

```diff
@@ -0,0 +1,180 @@
+
+    broadcast_event(ctx, "update", %{"fields" => updated_fields})
+
+    {:noreply, ctx}
+  end
+
+  defp update_options(ctx, df) do
+    df = String.to_atom(df)
+    options = ctx.assigns.options[df]
+    broadcast_event(ctx, "set_axis_options", %{"options" => options})
+  end
+
+  defp to_updates(_fields, field, ""), do: %{field => nil}
+
+  defp to_updates(_fields, field, value) when field in @as_int do
+    %{field => String.to_integer(value)}
+  end
+
+  defp to_updates(_fields, field, value) when field in @as_atom do
+    %{field => String.to_atom(value)}
+  end
+
+  defp to_updates(_fields, field, value), do: %{field => value}
+
+  defp vl_alias(%Macro.Env{aliases: [{vl_alias, VegaLite}]}), do: vl_alias
+  defp vl_alias(_), do: nil
+
+  @impl true
+  def to_attrs(ctx) do
+    ctx.assigns.fields
+    |> Map.put("vl_alias", ctx.assigns.vl_alias)
+  end
+
+  @impl true
+  def to_source(attrs) do
+    attrs
+    |> to_quoted()
+    |> Macro.to_string()
+    |> Code.format_string!()
+    |> IO.iodata_to_binary()
+  end
+
+  defp to_quoted(%{"data" => nil}), do: nil
+
+  defp to_quoted(attrs) do
+    data = attrs["data"]
+    module = if attrs["vl_alias"], do: attrs["vl_alias"], else: VegaLite
+    module = Module.split(module) |> hd() |> String.to_atom()
+
+    attrs =
+      attrs
+      |> Enum.map(fn {k, v} -> {String.to_atom(k), v} end)
+      |> Enum.into(%{})
+
+    [{_, root} | nodes] = [
+      new: %{name: :new, module: module, args: [[width: attrs.width, height: attrs.height]]},
+      data: %{name: :data_from_series, module: module, args: [Macro.var(data, nil)]},
+      mark: %{name: :mark, module: module, args: [attrs.chart]},
+      x: %{name: :encode_field, module: module, args: [attrs.x_axis, [type: attrs.x_axis_type]]},
+      y: %{name: :encode_field, module: module, args: [attrs.y_axis, [type: attrs.y_axis_type]]},
+      color: %{name: :encode_field, module: module, args: [attrs.color, [type: attrs.color_type]]}
+    ]
+
+    root = build_root(root)
+
+    nodes
+    |> Enum.map(&clean_node(&1))
+    |> Enum.reduce(root, &apply_node/2)
+  end
+
+  defp build_root(node) do
+    args =
+      node.args
+      |> Enum.at(0)
+      |> Enum.filter(&elem(&1, 1))
+      |> case do
+        [] -> []
+        opts -> [opts]
+      end
+
+    {{:., [], [{:__aliases__, [alias: false], [node.module]}, node.name]}, [], args}
+  end
+
+  def clean_node({field, node}) do
+    node = Map.replace!(node, :args, clean_args(node.args))
+    {field, node}
+  end
+
+  defp apply_node({_, %{args: nil}}, acc), do: acc
+  defp apply_node({_, %{args: []}}, acc), do: acc
+
+  defp apply_node({field, %{name: function, module: module, args: args}}, acc) do
+    ctx = [context: Elixir, import: Kernel]
+    args = if function == :encode_field, do: List.insert_at(args, 0, field), else: args
+
+    {:|>, ctx,
+     [
+       acc,
+       {{:., [], [{:__aliases__, [alias: false], [module]}, function]}, [], args}
+     ]}
```

Comment:
I think it can be something like this:

```suggestion
    args = if function == :encode_field, do: List.insert_at(args, 0, field), else: args
    
    quote do
      unquote(acc) |> unquote(module).unquote(function)(unquote_splicing(args))
    end
```

We could likely simplify other places with `quote` too, but it's gonna be easier to refactor when we have some tests for the code generation :) You can test the `to_source` function directly as we do [here](https://github.com/livebook-dev/kino/blob/ead0a6f42727eeec95c450b3bb5632c6ea9133e2/test/kino/smart_cell/sql_test.exs#L63-L173).